### PR TITLE
fix(trace-analytics): let the fold heal an undecodable row, and default the last two Map columns

### PR DESCRIPTION
**Pull request by 0xdeafcafe**
Two halves of the same defect, found chasing a `trace_processing` backlog on 2026-07-29.

An `ADD COLUMN` of a variable-size type without a DEFAULT leaves the column unmaterialised in every part written before the ALTER. A read then decodes a size header that was never written and asks for an absurd allocation:

```
Code: 173. Amount of memory requested to allocate is more than allowed:
(while reading column AnnotationIds) ... max_rows_to_read = 1
```

00014 fixed this once, 00057 fixed it again. Both wrote the rule down; neither stopped it hurting, because **the schema fix only ever addressed half the problem.**

## 1. Stop it wedging the pipeline

`trace_analytics`'s read-back caught this failure, logged it, and rethrew.

But the fold already owns a repair path. A **miss** refolds from `event_log` and writes a new version, whose higher `UpdatedAt` wins the dedup IN-tuple — so the damaged row simply stops being selected. One successful write supersedes it permanently.

Rethrowing denied the fold that path. The job died before the write, GroupQueue redelivered, the read hit the same bytes, and the group livelocked. On 2026-07-28 that wedged `trace_processing` for one tenant across many traces, at ~13,400 errors/hour peak.

This class of failure now reports a miss, so each affected trace heals on next touch.

**The narrowness is the safety property.** It requires BOTH `Code: 173` AND `while reading column`:

- code 173 alone is a genuine out-of-memory on a large-but-valid read, which a retry *can* fix — refolding it would discard a healthy row
- `while reading column` alone appears on unrelated read failures

Together they mean one column's stream is undecodable, which no retry repairs. A broad "read failed → refold" would turn a ClickHouse blip into a fleet-wide refold storm — the pattern `00061` blames for the 2026-07-23 TOO_MANY_PARTS outage. Four tests pin failures that must still throw: plain OOM, code 241, mid-deploy `UNKNOWN_IDENTIFIER` (ADR-066's deploy window), and transport errors.

Logged at warn rather than error — it is handled and self-healing, and `queryWindowed` already counts the `error` outcome on `clickhouse_windowed_read_total`, so the occurrence stays visible in metrics either way.

## 2. Close the last instance of the schema defect

00057 fixed seven Array columns and explicitly deferred `simulation_runs.RoleCosts` / `RoleLatencies` — Map rather than Array, not implicated in that incident, and a Map default needs its own type-checked literal.

Confirmed still outstanding on 2026-07-29 via `system.columns` across `clusterAllReplicas`: blank `default_expression` for both, on all three replicas, while every column 00057 touched reports `[]` on all three. `00064` gives them `DEFAULT map()`.

**After this, no ALTER-added Array or Map column is left without a DEFAULT.** Every other undefaulted variable-size column in `system.columns` (`stored_spans.*`, `metric_*.*`, `log_records.*`, `coding_agent_sessions.Steps`, `trace_analytics.{Attributes,Labels,Models}`, …) comes from its table's original CREATE, so it is materialised in every part and is not this defect.

Like 00057, metadata-only: no part rewritten, no mutation scheduled, idempotent.

## Verification

`pnpm typecheck` and `pnpm typecheck:tests` pass. 1,605 unit tests pass across the touched areas, including 5 new ones.

Production state was checked directly rather than inferred: `system.errors` shows code 173 last fired at **2026-07-28 21:48** on all three replicas, three minutes before the mutation that rewrote the affected parts. The specific damaged parts are gone. This PR is about the next occurrence, not the last one.

## Not in this PR

Two live findings from the same investigation, both separate from this defect:

- **`langwatch-workers` is crash-looping — ~47 restarts in the last hour**, `reason=Error` (not OOMKilled; memory sits at ~1.2 GiB). This is the actual current cause of the backlog: restarts kill in-flight jobs, which is why `spanStorage` shows 81 retries/min and why ClickHouse logs `Code 210 Broken pipe` (client vanishing mid-write). Needs its own diagnosis.
- **~50% of `trace_analytics` read-backs fall back to `unwindowed`** (108/min hit vs 106/min unwindowed). Per the ADR-068 note in that file, that ratio is the window-fit health signal, and on this table an unwindowed read is a tenant-wide dedup scan. A real throughput cost, independent of these errors.

## Summary by CodeRabbit

* **Bug Fixes**
  * Improved resilience when reading traces with specific ClickHouse unreadable-column failures, allowing affected records to be refolded instead of failing.
  * Preserved clear error reporting for other ClickHouse failure types (including schema drift, transport/network, and memory-related issues).
  * Prevented read-time failures for simulation run role cost and latency map data by applying safe default map values.

* **Tests**
  * Added coverage for corrupted data handling across multiple ClickHouse error scenarios.

### CODE REVIEW

File: langwatch/src/server/clickhouse/migrations/00064_simulation_runs_map_defaults.sql [hunk 1]
```diff
@@ -0,0 +1,66 @@
...
+-- +goose StatementEnd
+
+-- +goose ENVSUB OFF
+
+-- +goose Down
+-- IRREVERSIBLE: rolling back reinstates the defect. Dropping the DEFAULT puts
+-- parts written before 00008 back to decoding a size header that was never
+-- written, which is the read-time OOM this migration exists to stop. The Down
+-- migration is therefore commented out to prevent accidental data loss. To roll
+-- back, uncomment and run manually.
```
[the remark this answers is not in the document]
Comment (0xdeafcafe) [reply langwatch/src/server/clickhouse/migrations/00064_simulation_runs_map_defaults.sql]: Fixed in 7586fa1a2e. You are right that the old wording was wrong about the mechanism: dropping a `DEFAULT` is metadata-only and destroys nothing. The comment now says what the actual risk is — reinstating the read-time OOM on parts written before 00008 — and notes that the cost lands on the next read of an old part, not on the ALTER itself.

**Comment by langwatch-agent** [comment]
**Human Review Brief** (attention map, not a review)

**Mode:** Deep Review (small diff, live production defect)
**Linked issue:** none declared. Found chasing a real `trace_processing` backlog on 2026-07-29, so there is an active incident behind it.

**The defect is worth understanding before reading the code:** `ADD COLUMN` of a variable-size type with no DEFAULT leaves the column unmaterialised in every part written before the ALTER. A read then decodes a size header that was never written and requests an absurd allocation. That is a data-shape bug in ClickHouse, not an application bug, so the fix has to be right at both layers.

**Top inspection targets**
1. `Must Check` - the DEFAULT actually materialises for existing parts, or the read path tolerates unmaterialised columns for all of them. Adding a DEFAULT going forward does not retroactively fix parts already on disk; confirm which mechanism is relied on and that it covers the existing backlog.
2. `Must Check` - "let the fold heal an undecodable row" is the dangerous half. Healing must not mean swallowing. Confirm a row that cannot be decoded is counted and observable rather than silently dropped, or analytics quietly lose data and nobody learns.
3. `Ask Author` - whether other tables have had a variable-size `ADD COLUMN` without DEFAULT. This is a repeatable migration mistake, and the same read path will hit the same allocation error elsewhere.

**Production risk (scariest first):** silent analytics data loss if healing drops rows without a counter. Second: the memory-allocation error continuing on other tables with the same migration pattern.

**Questions for the author (max 3)**
1. Are already-written parts fixed, or only new ones?
2. Is a healed row counted in a metric?
3. Do other tables share this ADD COLUMN pattern?

File: langwatch/src/server/app-layer/traces/repositories/trace-analytics.clickhouse.repository.ts [hunk 2]
```diff
@@ -309,6 +345,30 @@ export class TraceAnalyticsClickHouseRepository
...
+      //
+      // Logged at warn, not error: this is a handled, self-healing path. The
+      // `error` outcome on clickhouse_windowed_read_total is already counted by
+      // queryWindowed, so the occurrence stays visible in metrics either way.
+      if (isUnreadableColumnError(error)) {
+        logger.warn(
+          { tenantId, traceId, error },
+          "Unreadable column in trace analytics row — refolding from event log",
+        );
+        return null;
```
Comment (langwatch-agent) [inline langwatch/src/server/app-layer/traces/repositories/trace-analytics.clickhouse.repository.ts:369]: Returning `null` makes this indistinguishable from a genuinely absent row. The executor therefore receives `miss: "absent"`, not `"undecodable"`, so its fail-closed guard does not apply. In the trace fold, the paged refold returns `null` when `eventLoaderUpToPaged` returns no rows; execution then falls through to `init() + delivered event` and stores it with a higher `UpdatedAt`. That silently replaces the complete but unreadable row with partial analytics whenever the event-log read is empty/lagging (or its history has aged out). The `absent` refold metric is also indistinguishable from a legitimate first event.

Carry this condition through the repository/store as `miss: "undecodable"` (or an equivalent typed signal), so an empty refold throws and retries rather than committing a partial replacement. Please add an end-to-end executor/store regression test proving an unreadable read plus an empty history does not call `store`.

**Review by langwatch-agent — CHANGES_REQUESTED** [review]
## Review

### Blocking
- The corrupt-row path is not carried through as an undecodable miss. If the event-log refold has no readable history, the executor treats the corrupt row as a normal new aggregate and writes only the currently delivered event, silently superseding the complete analytics row.

### Validated
- `DEFAULT map()` is valid for the existing physical type and applies at read time to pre-ALTER parts without scheduling a mutation (verified on disposable ClickHouse 26.7).
- The migration-history sweep supports the stated scope: the two `simulation_runs` maps are the remaining live columns in the `ADD COLUMN Array/Map without DEFAULT` pattern; prior instances are already defaulted or dropped.

I did not find another material issue in the migration itself.
